core/trino-main/src/main/java/io/trino/sql/planner/QueryPlanner.java

```diff
@@ -584,10 +587,32 @@ public UpdateNode plan(Update node)
         PlanAndMappings planAndMappings = coerce(builder, orderedColumnValues, analysis, idAllocator, symbolAllocator, typeCoercion);
         builder = planAndMappings.getSubPlan();
 
-        ImmutableList.Builder<Symbol> updatedColumnValuesBuilder = ImmutableList.builder();
-        orderedColumnValues.forEach(columnValue -> updatedColumnValuesBuilder.add(planAndMappings.get(columnValue)));
+        TableMetadata tableMetadata = metadata.getTableMetadata(session, handle);
+        ImmutableMap<String, Integer> updatedColumnNameToIndex = IntStream.range(0, updatedColumnNames.size())
+                .boxed()
+                .collect(toImmutableMap(updatedColumnNames::get, i -> i));
+        ImmutableList.Builder<Symbol> updatedNonNullableColumnsValuesBuilder = ImmutableList.builder();
+        Assignments.Builder assignmentsBuilder = Assignments.builder();
+        assignmentsBuilder.putIdentities(builder.getRoot().getOutputSymbols());
+
+        for (ColumnMetadata column : tableMetadata.getColumns()) {
```

Comment:
It feels the iteration would be simpler if you just iterate in parallel over `orderedColumValues` and `targetColumnNames`. IIUC those have same length so you can just do the iteration using integer index from `0..len(orderedColumValues)`.
The you would just look at columns you are interested in and look up for extra information (about nullability) by looking into `tableMetadata`. Currently you are iterating over all the columns and skipping once you are not interested in. This is counterintutive and requires using `updatedColumnNameToIndex`.

Reply:
So actually i thought about that but how would you look into tableMetadata for specific column ?
There is `getColumn` method but its implementation doesn't seem very efficient:
```
public ColumnMetadata getColumn(String name)
    {
        return getColumns().stream()
                .filter(columnMetadata -> columnMetadata.getName().equals(name))
                .findFirst()
                .orElseThrow(() -> new IllegalArgumentException("Invalid column name: " + name));
    }
```
We may end up going over that column list many many times. WDYT @losipiuk

Reply:
Can you just cache it as a name->CM map?